Add payment address parsing and tests for standard witness scripts

thecodefactory opened the pull request:
Add payment address parsing and tests for standard witness scripts, including address formats specified in BIP 142 and BIP 173 (v0 only).

Resolves #844

thecodefactory commented:
Interesting build error (fails on clang, good on gcc).  Testing clang locally to resolve.

thecodefactory commented:
I think this is in decent shape for merging (in terms of correctness), although I expect the interface may change some after @evoskuil has a chance to simplify it.

I have a separate (incoming, but nearly complete) PR for explorer that really makes this usable (including ec-to-witness and witness-to-key commands).

thecodefactory commented on include/bitcoin/system/wallet/witness_address.hpp:
```diff
@@ -0,0 +1,175 @@
+    typedef std::vector<witness_address> list;
+    typedef std::shared_ptr<witness_address> ptr;
+
+    /// Public for unit testing.
```
Ah, good to know, thanks!

evoskuil replied:
This sort of thing is really a pain in the ass in C++. I miss the use of [moq](https://github.com/moq/moq) and [friend assemblies](https://docs.microsoft.com/en-us/dotnet/standard/assembly/friend-assemblies?tabs=csharp), no need to expose internals just to properly unit test.

thecodefactory commented on src/wallet/witness_address.cpp:
```diff
@@ -0,0 +1,544 @@
+        const uint8_t witness_version = bech32_decoded.payload.front();
+
+        // Checks specific to witness version 0.
+        if (witness_version == 0 && (address.size() < 42 || address.size() > 62))
```
:+1:  Agreed, these BIP specified lengths should be better handled.

evoskuil replied:
Yes, IAW style guide, give each magic number a name.

evoskuil commented on include/bitcoin/system/wallet/witness_address.hpp:
```diff
@@ -0,0 +1,169 @@
+
+static BC_CONSTEXPR uint8_t padding = 0x00;
+static BC_CONSTEXPR size_t witness_p2wpkh_size = 3u + short_hash_size + checksum_size;
+static BC_CONSTEXPR size_t witness_p2wsh_size = 3u + hash_size + checksum_size;
```
Declare `witness_prefix_size = 3u`. Should update other classes as well so that magic numbers are self-documenting.

evoskuil commented on include/bitcoin/system/wallet/witness_address.hpp:
```diff
@@ -0,0 +1,169 @@
+        testnet_p2wsh,
+
+        unknown
+    };
```
For consistency you should expose the "base58" values as to `static const` and as named in `payment_address()`. You should drop the `enum` as the remaining are redundant and reduces flexibility. Undeclared prefixes must be allowed for alts.

evoskuil commented on include/bitcoin/system/wallet/witness_address.hpp:
```diff
@@ -0,0 +1,169 @@
+        encoding out_type=encoding::unknown);
+    witness_address(const std::string& address,
+        encoding out_type=encoding::unknown);
+    witness_address(short_hash&& hash, encoding out_type=encoding::unknown,
```
Avoid using the `out_` prefix for an in parameter. Just name the parameter `version` as it matches your `enum` comment and `payment_address`.

evoskuil commented on include/bitcoin/system/wallet/witness_address.hpp:
```diff
@@ -0,0 +1,169 @@
+    operator bool() const;
+
+    /// Serializer.
+    std::string bech32(const std::string& prefix=mainnet_prefix) const;
```
This method should not be provided as it allows one to change the prefix for an already constructed address. The address object should contain all information upon construct. This also implies there is no use for the `unknown` version value.

evoskuil commented on include/bitcoin/system/wallet/witness_address.hpp:
```diff
@@ -0,0 +1,169 @@
+    witness_address(witness_address&& other);
+    witness_address(const witness_address& other);
+    witness_address(const witness_p2wpkh& decoded,
+        encoding out_type=encoding::unknown);
```
An address object should never have an unknown version value. The object encompasses a complete address.

evoskuil commented on test/wallet/witness_address.cpp:
```diff
@@ -0,0 +1,213 @@
+        BOOST_REQUIRE(decode_base32(decoded, test.address));
+        uint8_t witness_version = decoded.payload.front();
+        // https://github.com/sipa/bech32/blob/master/ref/javascript/tests.js#L24
+        if (witness_version != 0)
```
Avoid conditional tests (test style guide).

evoskuil commented on test/wallet/witness_address.cpp:
```diff
@@ -0,0 +1,213 @@
+            witness_version += 0x50;
+
+        witness_address_accessor converter;
+        const auto witness_program = converter.convert_bits(5, 8, false,
```
Avoid wrapping test lines (test style guide)

evoskuil replied:
This is an address, not a converter, so just refer to it as `address`.

evoskuil replied:
Avoid magic number literals.

evoskuil commented on test/wallet/witness_address.cpp:
```diff
@@ -0,0 +1,213 @@
+{
+    for (const auto& test: native_witness_tests)
+    {
+        if (test.encoding == witness_address::encoding::mainnet_base58_p2wpkh ||
```
Avoid conditional tests.

evoskuil commented on test/wallet/witness_address.cpp:
```diff
@@ -0,0 +1,213 @@
+            witness_program.size());
+
+        data_chunk result = { witness_version, witness_program_size };
+        extend_data(result, witness_program);
```
I think there's an overload for extending using a set of values, so this could be simplified.

evoskuil commented on test/wallet/witness_address.cpp:
```diff
@@ -0,0 +1,213 @@
+            decoded.payload, 1);
+        BOOST_REQUIRE(witness_program.size() > 1 &&
+                      witness_program.size() < 41);
+        const auto witness_program_size = static_cast<uint8_t>(
```
Move up so and use in test macros so this is not computed 3 times.

evoskuil commented on test/wallet/witness_address.cpp:
```diff
@@ -0,0 +1,213 @@
+
+typedef std::vector<test_address> test_address_list;
+
+test_address_list witness_address_tests =
```
Use `const`.

evoskuil commented on test/wallet/witness_address.cpp:
```diff
@@ -0,0 +1,213 @@
+    { "BC1SW50QA3JX3S", "6002751e" },
+    { "bc1zw508d6qejxtdg4y5r3zarvaryvg6kdaj", "5210751e76e8199196d454941c45d1b3a323" },
+    { "tb1qqqqqp399et2xygdj5xreqhjjvcmzhxw4aywxecjdzew6hylgvsesrxh6hy", "0020000000c4a5cad46221b2a187905e5266362b99d5e91c6ce24d165dab93e86433" }
+};
```
As a constant declaration these can be nested (style).

evoskuil commented on test/wallet/witness_address.cpp:
```diff
@@ -0,0 +1,213 @@
+        const auto witness_program = converter.convert_bits(5, 8, false,
+            decoded.payload, 1);
+        BOOST_REQUIRE(witness_program.size() > 1 &&
+                      witness_program.size() < 41);
```
Avoid magic number literals.

evoskuil commented on test/wallet/witness_address.cpp:
```diff
@@ -0,0 +1,213 @@
+        uint8_t witness_version = decoded.payload.front();
+        // https://github.com/sipa/bech32/blob/master/ref/javascript/tests.js#L24
+        if (witness_version != 0)
+            witness_version += 0x50;
```
Avoid magic number literals.

evoskuil commented on test/wallet/witness_address.cpp:
```diff
@@ -0,0 +1,213 @@
+        bool pad, const data_chunk& in, size_t in_offset)
+    {
+        return witness_address::convert_bits(
+            from_bits, to_bits, pad, in, in_offset);
```
Normally want to break as close as possible to 80 chars, vs. nest args.

evoskuil commented on test/wallet/witness_address.cpp:
```diff
@@ -0,0 +1,213 @@
+    uint8_t witness_version;
+};
+
+typedef std::vector<test_vector> test_list;
```
No need for a `typedef` here (or even an explicit type?).

evoskuil commented on src/wallet/witness_address.cpp:
```diff
@@ -0,0 +1,632 @@
+}
+
+witness_address::witness_address(const data_chunk& data, encoding out_type)
+  : witness_address(from_data(data, out_type))
```
Why is arbitrary data allowed?

evoskuil commented on src/wallet/witness_address.cpp:
```diff
@@ -0,0 +1,632 @@
+static witness_address::encoding get_bech32_encoding(const std::string& address,
+    size_t hash_length)
+{
+    const auto prefix = address.substr(0, 2);
```
Safe for < 2 char string?

evoskuil replied:
Avoid magic literals.

evoskuil commented on src/wallet/witness_address.cpp:
```diff
@@ -0,0 +1,632 @@
+    else if (prefix == "tb")
+        return hash_length == short_hash_size ?
+            witness_address::encoding::testnet_p2wpkh :
+            witness_address::encoding::testnet_p2wsh;
```
How would this work for alts?

evoskuil commented on src/wallet/witness_address.cpp:
```diff
@@ -0,0 +1,632 @@
+        return hash_length == short_hash_size ?
+            witness_address::encoding::mainnet_p2wpkh :
+            witness_address::encoding::mainnet_p2wsh;
+    else if (prefix == "tb")
```
Avoid magic literals.

evoskuil commented on src/formats/base_32.cpp:
```diff
@@ -122,8 +122,8 @@ data_chunk checksum(const base32& value)
 
-    for (size_t index = 0; index < checksum.size(); ++index)
-        checksum[index] = (modified >> (5 * (5 - index))) & 31;
+    for (size_t index = 0; index < checksum_size; ++index)
```
Use `auto` (no need for more than int here).

thecodefactory replied:
After considering, I left this (and a different instance, below) as-is to avoid compiler warnings:

```
src/formats/base_32.cpp: In function ‘uint32_t libbitcoin::system::polymod(const data_chunk&)’:
src/formats/base_32.cpp:102:36: warning: comparison of integer expressions of different signedness: ‘int’ and ‘size_t’ {aka ‘long unsigned int’} [-Wsign-compare]
         for (auto index = 0; index < bit_group_size; ++index)
                              ~~~~~~^~~~~~~~~~~~~~~~
src/formats/base_32.cpp: In function ‘libbitcoin::system::data_chunk libbitcoin::system::checksum(const libbitcoin::system::base32&)’:
src/formats/base_32.cpp:127:32: warning: comparison of integer expressions of different signedness: ‘int’ and ‘size_t’ {aka ‘long unsigned int’} [-Wsign-compare]
     for (auto index = 0; index < checksum_size; ++index)
```
Alternatively, could use int for the constants ...

evoskuil commented on src/formats/base_32.cpp:
```diff
@@ -122,8 +122,8 @@ data_chunk checksum(const base32& value)
-    for (size_t index = 0; index < checksum.size(); ++index)
-        checksum[index] = (modified >> (5 * (5 - index))) & 31;
+    for (size_t index = 0; index < checksum_size; ++index)
+        checksum[index] = (modified >> 5 * (5 - index)) & 31;
```
Avoid magic literals.

thecodefactory commented:
Updated, but I need to track down the clang presented bugs next ...

thecodefactory commented:
Updated.

@evoskuil The use of `version` is all around much better than the previous encoding stuff when it's well defined and used properly.  This part still seems problematic to me:

https://github.com/libbitcoin/libbitcoin-system/pull/1118/files#diff-e73f757b65a6660336e35268a78db663R48

In short, I'm using some `fake` version codes simply to distinguish the cases where a witness address is constructed from a script or public key and we need to distinguish between the identity as bech32 p2wpkh or p2wsh encoding.  It could probably be replaced with something as simple as an additional `bool` on construction, but wanted your input on it.  `Version` is currently re-used in this case because it's already there (and mimics the actually defined versions for the base58 equivalents), but as it is now, it does mean that we maintain fake version mappings that may eventually conflict with actual versions (perhaps in the alt space).  If we did add a bool for that determination, the `version` is ignored for cases where we don't use the base58 encoding.  The other downside of not using the `version` mapping is that it makes more conditional logic required for things like this:

https://github.com/libbitcoin/libbitcoin-system/pull/1118/files#diff-e73f757b65a6660336e35268a78db663R299

thecodefactory commented:
Just pushed one more minor update/clean-up ...

evoskuil commented on src/wallet/witness_address.cpp:
```diff
@@ -0,0 +1,550 @@
+
+/// static
+void witness_address::to_witness(witness_p2wpkh& out, uint8_t version,
+    uint8_t witness_version, short_hash hash)
```
Avoid parameter copies by always using `const short_hash& hash` idiom). When a copy is required, make it internally for the sake of consistency.

evoskuil replied:
See other instances.

evoskuil commented on src/wallet/witness_address.cpp:
```diff
@@ -0,0 +1,550 @@
+    return valid_ == other.valid_ && version_ == other.version_ &&
+        prefix_ == other.prefix_ &&
+        witness_version_ == other.witness_version_ && hash_ == other.hash_ &&
+        witness_hash_ == other.witness_hash_;
```
This is the case where I prefer stacking (a long sequence of `&&` or `||` operations), an exception to the stacking style rule. 
```cpp
    return valid_ == other.valid_
        && version_ == other.version_
        && prefix_ == other.prefix_
        && witness_version_ == other.witness_version_
        && hash_ == other.hash_
        && witness_hash_ == other.witness_hash_;
```

evoskuil commented on test/wallet/witness_address.cpp:
```diff
@@ -0,0 +1,265 @@
+BOOST_AUTO_TEST_CASE(witness_address__construct__from_ec_public__valid_expected)
+{
+    const witness_address address(ec_public(BIP142_PUBLIC_KEY),
+        witness_address::mainnet_base58_p2wpkh);
```
Avoid test line wrap, lots more below. Can use intermediate vars to break up (perf not a material consideration).

evoskuil commented on test/wallet/witness_address.cpp:
```diff
@@ -0,0 +1,265 @@
+    for (const auto& test: witness_address_nonzero_version_tests)
+    {
+        base32 decoded;
+        BOOST_REQUIRE(decode_base32(decoded, test.address));
```
I tend to group the statements such that `BOOST_REQUIRE(...)` (or a set of them) terminates the section, leaving a blank line after, then on to the next group.

evoskuil commented:
> I'm using some fake version codes simply to distinguish the cases where a witness address is constructed from a script or public key and we need to distinguish between the identity as bech32 p2wpkh or p2wsh encoding.

Okay, so there is an enumeration of possible methods of encoding, that is independent of the other information, but currently it is being inferred from that information? This calls for an independent enumeration, but as you suggest, since there are only two valid options as per the specification, it could be Boolean. I would refer to it as something other than "version" or "encoding", given that those are overloaded in the same context. The current distinction is between "public-key" and "script" style of hash/address. Maybe a class enumeration named `address_type` with two options, `key` and `script`. Allowing an invalid option, such as `unknown` will only unnecessarily complicate the code, given that the class enum is a constraint.

thecodefactory commented on src/wallet/witness_address.cpp:
```diff
@@ -0,0 +1,550 @@
+    if (secret)
+        return { secret.to_public(), version, prefix };
+
+    return {};
```
For almost all of the ternary return cases, we get this error if attempted:

```
error: initializer list cannot be used on the right hand side of operator '?'
...
error: initializer list cannot be used on the right hand side of operator ':'
```

It's probably not worth first assigning, then using the ternary return, as it doesn't simplify/shorten things.

evoskuil replied:
You need to explicitly specify the type and it will work.

evoskuil replied:
Not a huge deal tho.

thecodefactory commented:
>> I'm using some fake version codes simply to distinguish the cases where a witness address is constructed from a script or public key and we need to distinguish between the identity as bech32 p2wpkh or p2wsh encoding.

> Okay, so there is an enumeration of possible methods of "encoding", that is independent of the other information, but currently it is being inferred from that information? This calls for an independent enumeration, but as you suggest, since there are only two valid options as per the specification, it could be Boolean. I would refer to it as something other than "version" or "encoding", given that those are overloaded in the same context. The current distinction is between "public-key" and "script" style of hash/address. Maybe a class enumeration named address_format with two options, public_key and script. Allowing an invalid option, such as unknown will only unnecessarily complicate the code, given that the class enum is a constraint. Then using format for the parameter names.

Sort of.  When the witness address is constructed from hash/public/key/script, we want to know what will come out when `.encoded()` is called (for example, base58 p2wpkh, base58 p2wsh, bech32 p2wpkh, or bech32 p2wsh).  Construction takes that data, a version (required in the base58 cases to distinguish between p2wpkh and p2wsh, and re-used from the `payment_address` base), a witness version, and a prefix (unused for base58, but required for bech32).  What we know with those parameters for base58 encodings is everything required, and we have an extra unused prefix.  What we don't explicitly know with those parameters is whether the encoding should be bech32 p2wpkh, or bech32 p2wsh.  That's where we're re-using the version field (since it's only used in the base58 encoding cases otherwise).  We could add a new field, that simply specifies bech32 p2wpkh or bech32 p2wsh (and/or possibly mandates the use of base58 instead).

IFF we did that, in the base58 cases, now we ignore the prefix and the new encoding flag (unused).  And similarly in the bech32 cases, we ignore the version parameter passed in (only used to specify a base58 version).  This gets a little tricky because now we don't explicitly know which to ignore (the base58 version, or the new encoding flag, perhaps without some other piece of information).

Does this make sense?  Re-using the version (as we do now) appears to be the most concise, but it's confusing.

At least conceptually, what we may want is to split this out some more so that we have separate classes (naming is bad, just explicit for example):

```
base58_witness_address(hash/script/public_key, version, witness_version);

bech32_witness_address(hash/script/public_key, witness_version,
    prefix, p2wpkh_or_p2wsh_flag);
```

_Where hash/script/public_key means we'd have constructors for each of those (in addition to the other standard variants)_

The alternative is cramming it all into something like:

```
witness_address(hash/script/public_key, version, witness_version,
    prefix, p2wpkh_or_p2wsh_flag);
```

Which is what happens if we add a new flag/format field.

evoskuil commented:
Why are we attempting to support distinct base58 and bech32 encodings?

thecodefactory commented:
> Why are we attempting to support distinct base58 and bech32 encodings?

This implementation supports both BIP142 ("Address Format for Segregated Witness") and BIP173 ("Base32 address format for native v0-16 witness outputs").

evoskuil commented:
Overloading leads to complexity.

thecodefactory commented:
> Overloading leads to complexity.

Agreed, thus the design issue I'm fretting over :-(

But I also think splitting them out completely (like the poorly named examples above) is not quite right either.  I'm open to ideas/suggestions here.  I know there's a very workable middle ground somewhere, but I'm not seeing it just yet.

evoskuil commented:
[BIP142](https://github.com/bitcoin/bips/blob/master/bip-0142.mediawiki) is status: withdrawn.

thecodefactory commented:
> [BIP142](https://github.com/bitcoin/bips/blob/master/bip-0142.mediawiki) is status: withdrawn.

Wow, thanks, didn't realize!  If the answer is as simple as stripping out BIP142 support, I think I've got a clear resolution path here.

thecodefactory commented:
Updated, but will remove BIP142 support next to help simplify the interface.

thecodefactory commented:
Dropped BIP142 support and added an `address_format` field.  I think this is much less confusing than the previous code trying to support both base58 and bech32 addresses all in one.

evoskuil commented:
`magic_numbers[index]` - creative yet ineffective. The point of naming magic numbers is to eliminate the magic. This is still magic. ;)

evoskuil commented on src/chain/script.cpp:
```diff
@@ -1171,6 +1177,24 @@ operation::list script::to_pay_multisig_pattern(uint8_t signatures,
 
+operation::list script::to_pay_witness_key_hash_pattern(const short_hash& hash)
+{
+    return operation::list
```
`operation::list` may be dropped here as it's the return type (more below).

evoskuil commented on include/bitcoin/system/wallet/witness_address.hpp:
```diff
@@ -0,0 +1,142 @@
+    enum class address_format: uint8_t
+    {
+        p2wpkh,
+        p2wsh
```
Style: acronyms :/. Could you review against existing similar terms in `-system` and update for consistency?

evoskuil commented on src/formats/base_32.cpp:
```diff
@@ -91,13 +93,15 @@ uint32_t polymod(const data_chunk& values)
         0x3b6a57b2, 0x26508e6d, 0x1ea119fa, 0x3d4233dd, 0x2a1462b3
```
The magic numbers have moved. But what are these?

evoskuil commented on src/wallet/witness_address.cpp:
```diff
@@ -0,0 +1,367 @@
+        to_chunk(witness_version_),
+        witness_hash_ == null_hash ? convert_bits(bech32_expanded_bit_size,
+            bech32_contracted_bit_size, true, to_chunk(hash_), 0) :
+                convert_bits(bech32_expanded_bit_size,
```
You could outdent this "else" portion one tab width.

evoskuil commented on src/wallet/witness_address.cpp:
```diff
@@ -0,0 +1,367 @@
+{
+    return format_ == address_format::p2wpkh ?
+        chain::script::to_pay_witness_key_hash_pattern(hash_) :
+        chain::script::to_pay_witness_script_hash_pattern(witness_hash_);
```
Notice the above-mentioned outdenting is done here (same logical level).

evoskuil commented on test/wallet/witness_address.cpp:
```diff
@@ -0,0 +1,180 @@
+            decoded.payload, 1);
+
+        const auto witness_program_size = static_cast<uint8_t>(witness_program.size());
+        BOOST_REQUIRE(witness_program_size > witness_v0_program_minimum_limit &&
```
Test style: unwrap or (better) split into two tests.

evoskuil commented on test/wallet/witness_address.cpp:
```diff
@@ -0,0 +1,180 @@
+        witness_version += op_version_gap;
+
+        witness_address_accessor converter;
+        const auto witness_program = converter.convert_bits(
```
Test style: unwrap.

evoskuil commented on test/wallet/witness_address.cpp:
```diff
@@ -0,0 +1,180 @@
+            decoded.payload, 1);
+
+        const auto witness_program_size = static_cast<uint8_t>(witness_program.size());
+        BOOST_REQUIRE(witness_program_size > witness_v0_program_minimum_limit &&
```
Test style: unwrap or split.

thecodefactory commented:
@evoskuil I believe this has been updated to address all current feedback.  If I missed something, let me know!

evoskuil commented on src/formats/base_32.cpp:
```diff
@@ -86,22 +88,32 @@ data_chunk expand(const std::string& prefix)
+    // simple integers. Generally 30-bit integers are used, where each
+    // bit corresponds to one coefficient of the polynomial.
+    //
+    // https://bitcoin.stackexchange.com/questions/74573/how-is-bech32-based-on-bch-codes
```
Drop "https://" to shorten line and eliminate hyperlink highlighting in VS (which is annoying).

evoskuil commented on src/wallet/witness_address.cpp:
```diff
@@ -0,0 +1,383 @@
+    address_format format, const std::string& prefix)
+{
+    return secret ? witness_address(secret.to_public(), format, prefix)
+        : witness_address();
```
witness_address{}

evoskuil commented on src/wallet/witness_address.cpp:
```diff
@@ -0,0 +1,383 @@
+    const uint8_t witness_version = 0;
+
+    return format == address_format::witness_pubkey_hash ?
+        witness_address(bitcoin_short_hash(data), format, witness_version, prefix) :
```
line length, maybe just use `version`.

evoskuil commented on src/wallet/witness_address.cpp:
```diff
@@ -0,0 +1,383 @@
+            convert_bits(bech32_expanded_bit_size, bech32_contracted_bit_size,
+                true, to_chunk(hash_), 0) :
+            convert_bits(bech32_expanded_bit_size, bech32_contracted_bit_size,
+                true, to_chunk(witness_hash_), 0)
```
Use named const for `0` parameter.

evoskuil commented on src/wallet/witness_address.cpp:
```diff
@@ -0,0 +1,383 @@
+
+        accumulator = ((accumulator << from_bits) | value) &
+            maximum_accumulator;
+        bit_count += from_bits;
```
Could move this up one line for slightly better appearance.
